### PR TITLE
V2: Convert from float64 to float32

### PR DESCRIPTION
**Pull request by shihchengli**
## Description
MPS framework doesn't support float64. Some attributes in a datapoint are saved as float64. Converting them to float32 can solve this issue.

## Relevant issues
#711 

## Checklist
- [ ] linted with flake8?
- [ ] (if appropriate) unit tests added?

### CODE REVIEW

File: chemprop/data/datasets.py
Comment (davidegraff) [inline chemprop/data/datasets.py:1]: getting rid of the `float` casts in this file can potentially lead to target arrays beings parsed in as `int`. Is there a reason that removing them is necessary?

**Review by davidegraff — COMMENTED** [review]
is there a reason changing the numpy precision of the raw data fixes the MPS errors? AFAICT all data is cast to `torch.float32` during `BatchMolGraph` preparation, so it shouldn't really matter what numpy dtype we use prior to passing data in to the modules.

File: chemprop/data/datasets.py
Comment (shihchengli) [reply chemprop/data/datasets.py:1]: No, I will add this back. Thanks for the explanation.

**Comment by shihchengli** [comment]
@davidegraff You're right! I noticed that the `weights` aren't saved as `torch.float32` in the `BatchMolGraph`. I'll make the necessary changes accordingly. Thanks for your insights!

**Comment by davidegraff** [comment]
I see- I don't really know why this would matter considering that the following code produces the right dtype:
```python
>>> torch.tensor([1., 1., 1.]).dtype
torch.float32
```
**But**, if a user relies on numpy to initialize weights then it can be a problem:
```python
>>> weights = np.random.rand(3)
>>> data = [MoleculeDatapoint(..., weight=w) for w in weights]
>>> torch.tensor([d.weight for d in data]).dtype
torch.float64
```
So really it seems that explicit dtype conversion can fix this for us. That said, I don't think it's a bad thing to cast all `MolGraph` attributes to single precision floating point number (e.g., `dtype=np.float32|"f4"`). It would save us about half the memory (because we only use half the number of bits to store the data) and nothing is lost anyway (our double precision data is _always_ cast to single precision prior to being fed into the model.

File: chemprop/data/collate.py [hunk 1]
```diff
@@ -88,7 +88,7 @@ def collate_batch(batch: Iterable[Datum]) -> TrainingBatch:
         None if V_ds[0] is None else torch.from_numpy(np.concatenate(V_ds)).float(),
         None if x_ds[0] is None else torch.from_numpy(np.array(x_ds)).float(),
         None if ys[0] is None else torch.from_numpy(np.array(ys)).float(),
-        torch.tensor(weights).unsqueeze(1),
+        torch.tensor(weights).float().unsqueeze(1),
```
Comment (davidegraff) [inline chemprop/data/collate.py:91]:
```suggestion
        torch.tensor(weights, dtype=torch.float).float().unsqueeze(1),
```

**Comment by davidegraff** [comment]
@shihchengli - your original changes (that I have since convinced you to revert) could actually benefit the repo. Would you mind going back to this original change of casting all `MolGraph` attributes to single precision (e.g., `dtype="f4"`)?

File: chemprop/featurizers/molgraph/molecule.py [hunk 2]
```diff
@@ -60,9 +60,9 @@ def __call__(
             )
 
         if n_atoms == 0:
-            V = np.zeros((1, self.atom_fdim))
+            V = np.zeros((1, self.atom_fdim)).astype(np.float32)
```
Comment (davidegraff) [inline chemprop/featurizers/molgraph/molecule.py:63]:
```suggestion
            V = np.zeros((1, self.atom_fdim), dtype=np.single)
```
equivalently, `dtype="f4"`

File: chemprop/featurizers/molgraph/molecule.py [hunk 3]
```diff
@@ -60,9 +60,9 @@ def __call__(
...
         else:
-            V = np.array([self.atom_featurizer(a) for a in mol.GetAtoms()])
+            V = np.array([self.atom_featurizer(a) for a in mol.GetAtoms()]).astype(np.float32)
```
Comment (davidegraff) [inline chemprop/featurizers/molgraph/molecule.py:65]:
```suggestion
            V = np.array([self.atom_featurizer(a) for a in mol.GetAtoms()], dtype=np.single)
```

File: chemprop/cli/utils/parsing.py [hunk 4]
```diff
@@ -51,14 +51,14 @@ def parse_csv(
...
 
     Y = df[target_cols]
-    weights = None if weight_col is None else df[weight_col].to_numpy()
+    weights = None if weight_col is None else df[weight_col].astype(np.float32).to_numpy()
 
     if bounded:
         lt_mask = Y.applymap(lambda x: "<" in x).to_numpy()
         gt_mask = Y.applymap(lambda x: ">" in x).to_numpy()
-        Y = Y.applymap(lambda x: x.strip("<").strip(">")).astype(float).to_numpy()
+        Y = Y.applymap(lambda x: x.strip("<").strip(">")).astype(np.float32).to_numpy()
```
Comment (davidegraff) [inline chemprop/cli/utils/parsing.py:59]:
```suggestion
        Y = Y.applymap(lambda x: x.strip("<").strip(">")).to_numpy(np.single)
```

File: chemprop/cli/utils/parsing.py [hunk 5]
```diff
@@ -51,14 +51,14 @@ def parse_csv(
         target_cols = list(set(df.columns) - set(input_cols) - set(ignore_cols or []))
 
     Y = df[target_cols]
-    weights = None if weight_col is None else df[weight_col].to_numpy()
+    weights = None if weight_col is None else df[weight_col].astype(np.float32).to_numpy()
```
Comment (davidegraff) [inline chemprop/cli/utils/parsing.py:54]:
```suggestion
    weights = None if weight_col is None else df[weight_col].to_numpy(np.single)
```

File: chemprop/cli/utils/parsing.py [hunk 6]
```diff
@@ -51,14 +51,14 @@ def parse_csv(
...
 
     if bounded:
         lt_mask = Y.applymap(lambda x: "<" in x).to_numpy()
         gt_mask = Y.applymap(lambda x: ">" in x).to_numpy()
-        Y = Y.applymap(lambda x: x.strip("<").strip(">")).astype(float).to_numpy()
+        Y = Y.applymap(lambda x: x.strip("<").strip(">")).astype(np.float32).to_numpy()
     else:
-        Y = Y.to_numpy()
+        Y = Y.astype(np.float32).to_numpy()
```
Comment (davidegraff) [inline chemprop/cli/utils/parsing.py:61]:
```suggestion
        Y = Y.to_numpy(np.single)
```

**Review by davidegraff — COMMENTED** [review]
note: If you prefer using the keyword syntax (i.e., `.to_numpy(dtyle=np.single)`), I have no strong preference for positionals.

File: chemprop/cli/utils/parsing.py [hunk 7]
```diff
@@ -148,7 +148,7 @@ def make_datapoints(
     else:
         N = len(smiss)
 
-    weights = np.ones(N) if weights is None else weights
+    weights = np.ones(N, dtype=np.float32) if weights is None else weights
```
Comment (shihchengli) [inline chemprop/cli/utils/parsing.py:151]:
```suggestion
    weights = np.ones(N, dtype=np.single) if weights is None else weights
```

File: chemprop/featurizers/molgraph/molecule.py [hunk 8]
```diff
@@ -78,7 +78,9 @@ def __call__(
 
                 x_e = self.bond_featurizer(bond)
                 if bond_features_extra is not None:
-                    x_e = np.concatenate((x_e, bond_features_extra[bond.GetIdx()]))
+                    x_e = np.concatenate(
+                        (x_e, bond_features_extra[bond.GetIdx()]), dtype=np.float32
```
Comment (shihchengli) [inline chemprop/featurizers/molgraph/molecule.py:82]:
```suggestion
                        (x_e, bond_features_extra[bond.GetIdx()]), dtype=np.single
```

**Review by davidegraff — APPROVED** [review]
👍

**Review by KnathanM — APPROVED** [review]

**Pull request merged**